File: src/leetcode/minimum-operations-to-make-the-array-increasing/minimum-operations-to-make-the-array-increasing.js
/**
 * Leetcode #1827
 * Difficulty: Easy
 * https://leetcode.com/problems/minimum-operations-to-make-the-array-increasing/
 */
/**
 * @param {number[]} nums
 * @return {number}
 */
export const minOperations = nums => {
  let ans = 0;

  for (let i = 1; i < nums.length; i += 1) {
    if (nums[i] <= nums[i - 1]) {
      const diff = Math.abs(nums[i] - nums[i - 1]);
      nums[i] = nums[i] + diff + 1;
      ans += diff + 1;
    }
  }

  return ans;
};

File: src/leetcode/minimum-operations-to-make-the-array-increasing/minimum-operations-to-make-the-array-increasing.rs
impl Solution {
    pub fn min_operations(nums: Vec<i32>) -> i32 {
        let mut ans = 0;
        let mut prev = 0;

        for num in nums {
            if prev >= num {
                ans += prev - num + 1;
                prev += 1;
            } else {
                prev = num;
            }
        }

        ans
    }
}


// impl Solution {
//     pub fn min_operations(nums: Vec<i32>) -> i32 {
//         let mut prev = nums[0];
//
//         nums.iter().skip(1).fold(0, |acc,num| {
//             if num - prev <= 0 {
//                 let diff = prev - num;
//                 prev = num + diff + 1;
//                 return acc + diff + 1
//             }
//
//
//             prev = *num;
//             acc
//         })
//     }
// }